compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/util/IrUtils.kt

```diff
@@ -515,11 +512,27 @@ val IrFunction.allTypeParameters: List<IrTypeParameter>
     else
         typeParameters
 
+
 fun IrMemberAccessExpression<*>.getTypeSubstitutionMap(irFunction: IrFunction): Map<IrTypeParameterSymbol, IrType> {
     val typeParameters = irFunction.allTypeParameters
-    return if (typeParameters.isEmpty()) emptyMap() else typeParameters.withIndex().associate {
+    val dispatchReceiverTypeArguments = (dispatchReceiver?.type as? IrSimpleType)?.arguments ?: emptyList()
+    if (typeParameters.isEmpty() && dispatchReceiverTypeArguments.isEmpty()) {
+        return emptyMap()
+    }
+
+    val result = mutableMapOf<IrTypeParameterSymbol, IrType>()
+    if (dispatchReceiverTypeArguments.isNotEmpty()) {
+        val parentTypeParameters = extractTypeParameters(irFunction.parentClassOrNull!!)
+        parentTypeParameters.withIndex().forEach { (index, typeParam) ->
+            dispatchReceiverTypeArguments[index].typeOrNull?.let {
```

Comment:
Can we make this `typeOrNull!!` or would it fail in some case?

Reply:
It will fail for instance if there is a star projection:

Java:
```
public class AT<G> {
  public void foo(G ...x) {
  }
}
```

Kotlin:
```
fun AT<*>.bar() {
  foo()
}
```

Reply:
Right... For some reason I though this would mean a star in function's type arguments, which would be prohibited.